# ML-03 — Frame Your Lane as an ML Task

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Alizawwaris974/Assignment-1---Flyrank-ML/blob/main/work/notebooks/w02_ml_task_framing.ipynb)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [1]:
import os, sys, subprocess

IN_COLAB = "google.colab" in sys.modules
REPO_URL = "https://github.com/flyrank-bih/flyrank-ml-internship-starter"
REPO_DIR = "flyrank-ml-internship-starter"

if IN_COLAB:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
    os.chdir(REPO_DIR)
elif os.path.basename(os.getcwd()) == "notebooks" and os.path.basename(os.path.dirname(os.getcwd())) == "work":
    os.chdir("../..")

import pandas as pd, numpy as np
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")
df["is_declining_label"] = df["trend_direction"].str.lower().eq("down").astype(int)
print("Working dir:", os.getcwd(), "| rows:", len(df))

Working dir: /content/flyrank-ml-internship-starter | rows: 30000


## 1. My lane as an ML task (type)

*Classification, clustering, ranking, or scoring — which one, and why?*

Task type: Scoring/ranking, built on top of binary classification. The underlying model predicts a probability (is_declining_label), but the actual deliverable isn't the 0/1 prediction itself — it's a ranked queue (top-K pages to review), evaluated with Precision@K. This matches an editor's real constraint: fixed review capacity, not "classify every page."

## 2. Target or proxy

*What would you predict? Where does that label come from — observed outcome or a defined rule?*

The target is a defined-rule proxy, not a directly observed business outcome: is_declining_label = (trend_direction == "down"). It's a reasonable stand-in for "this page needs attention," but it's derived from a trend bucket, not from an editor's actual judgment of what's worth fixing — worth remembering as a limitation.

In [2]:
print(df["is_declining_label"].value_counts(normalize=True).round(3))

# The hand-rule proxy score from notebook 02, for comparison
stale = (df["days_since_last_update"] >= 180).astype(int)
visible = (df["impressions_90d"] >= 500).astype(int)
df["hand_rule_score"] = stale * visible * df["impressions_90d"]
df[["content_id", "trend_direction", "is_declining_label", "hand_rule_score"]].head(10)

is_declining_label
1    0.542
0    0.458
Name: proportion, dtype: float64


,content_id,trend_direction,is_declining_label,hand_rule_score
0,content_304f48230142,down,1,0
1,content_a1fb4e703a9e,down,1,0
2,content_9aa793d4d895,down,1,0
3,content_331d6c4de07b,stable,0,0
4,content_d99b7a2d90ca,down,1,0
5,content_d4084a4bc775,down,1,0
6,content_9a34b442b552,down,1,0
7,content_a63219c6e95a,stable,0,0
8,content_5e6c160719bc,down,1,0
9,content_c27558df2b0c,down,1,0


## 3. Success metric

*One metric you can defend. What number means 'good'?*
Metric: Precision@50 — matches an editor's real review capacity (they can't review all 30,000 pages, only a fixed batch). Overall accuracy would be misleading here anyway — it hides how well the model does specifically on the pages someone will actually act on.

In [8]:
def precision_at_k(scores, labels, k):
    order = np.argsort(-np.asarray(scores))
    return np.asarray(labels)[order[:k]].mean()

y = df["is_declining_label"].values
for k in (20, 50):
    print(f"Hand rule Precision@{k}: {precision_at_k(df['hand_rule_score'], y, k):.3f}")

nonzero = (df["hand_rule_score"] > 0).sum()
print(f"\nCaveat: only {nonzero} of {len(df)} rows have a nonzero hand_rule_score.")
print("Precision@20/50 above are mostly determined by tie-breaking among score=0 rows, not real ranking.")

Hand rule Precision@20: 0.900
Hand rule Precision@50: 0.680

Caveat: only 17 of 30000 rows have a nonzero hand_rule_score.
Precision@20/50 above are mostly determined by tie-breaking among score=0 rows, not real ranking.


## 4. The unit of analysis, as a real dataframe

*Load your lane's slice and show it: one row = one what?*

In [4]:
unit_of_analysis = df[[
    "content_id", "client_id", "trend_direction", "is_declining_label",
    "days_since_last_update", "impressions_90d", "hand_rule_score"
]]
print(f"{len(unit_of_analysis)} rows -> one row per page")
unit_of_analysis.head(5)

30000 rows -> one row per page


,content_id,client_id,trend_direction,is_declining_label,days_since_last_update,impressions_90d,hand_rule_score
0,content_304f48230142,client_f369cb89fc,down,1,20,3803,0
1,content_a1fb4e703a9e,client_4e07408562,down,1,25,15320,0
2,content_9aa793d4d895,client_7f2253d7e2,down,1,20,12581,0
3,content_331d6c4de07b,client_19581e27de,stable,0,22,11751,0
4,content_d99b7a2d90ca,client_3fdba35f04,down,1,14,19140,0


## 5. Why ML beats a fixed rule here

*What makes the pattern too messy for an if-statement?*

Neither the hand rule nor the decision tree gives a trustworthy Precision@50 at this k, and going deeper (depth 2 → 4) doesn't fix it:

Hand rule: only 17/30,000 rows have a real score; the rest of any "top-K" list is arbitrary tie-breaking.
Tree, depth 2: top leaf has 17,816 rows (true decline rate 65.1%); Precision@50 (0.600) is just noise around that average.
Tree, depth 4: top leaf shrinks to 10,442 rows (true decline rate 71.6%); Precision@50 (0.680) is still just noise around that average — barely different in structure, even though the raw number moved.

The real lesson: a decision tree can only rank pages between leaves, never within one. Depth controls how many leaves you get, but as long as any single leaf is larger than your target k, the model has no way to distinguish the "best 50" from the rest of that leaf — you'd need either a leaf smaller than 50 (which risks overfitting to a tiny group) or a fundamentally different kind of model that assigns a continuous, mostly-unique score per row (e.g., logistic regression, gradient boosting, or a random forest averaging many trees) rather than one shared score per leaf.

This reframes the "why ML beats a fixed rule" argument for Lane 2: it's not that a shallow tree already beats the hand rule — it doesn't, cleanly. It's that neither a hard-coded threshold nor a too-shallow tree has enough resolution to support a k=50 decision, and the fix is a model built for fine-grained per-row scoring, not just "any ML model."

In [9]:
from sklearn.tree import DecisionTreeClassifier

features = ["content_age_days", "days_since_last_update", "impressions_90d", "avg_position", "ctr", "word_count"]
X = df[features].replace([np.inf, -np.inf], np.nan).fillna(0)

print("Distinct hand_rule_score values:", df["hand_rule_score"].nunique(), "(rest are ties at 0)")

for depth in (2, 4):
    t = DecisionTreeClassifier(max_depth=depth, class_weight="balanced", random_state=42).fit(X, y)
    score = t.predict_proba(X)[:, 1]
    n_distinct = len(np.unique(score))
    p50 = precision_at_k(score, y, 50)
    print(f"depth={depth}: {n_distinct} distinct scores, Precision@50 = {p50:.3f}")

    order = np.argsort(-score)
    top_leaf_score = score[order[0]]
    same_leaf_in_top50 = (np.round(score[order[:50]], 6) == round(top_leaf_score, 6)).sum()
    print(f"  -> {same_leaf_in_top50}/50 of the top-50 share the single highest score (tie size at the top)")

Distinct hand_rule_score values: 18 (rest are ties at 0)
depth=2: 4 distinct scores, Precision@50 = 0.600
  -> 50/50 of the top-50 share the single highest score (tie size at the top)
depth=4: 16 distinct scores, Precision@50 = 0.680
  -> 50/50 of the top-50 share the single highest score (tie size at the top)


In [10]:
from sklearn.tree import DecisionTreeClassifier

features = ["content_age_days", "days_since_last_update", "impressions_90d", "avg_position", "ctr", "word_count"]
X = df[features].replace([np.inf, -np.inf], np.nan).fillna(0)

print("Distinct hand_rule_score values:", df["hand_rule_score"].nunique(), "(rest are ties at 0)")

for depth in (2, 4):
    t = DecisionTreeClassifier(max_depth=depth, class_weight="balanced", random_state=42).fit(X, y)
    score = t.predict_proba(X)[:, 1]
    n_distinct = len(np.unique(score))
    p50 = precision_at_k(score, y, 50)

    top_score = score.max()
    mask = np.round(score, 6) == round(top_score, 6)
    print(f"depth={depth}: {n_distinct} distinct scores, Precision@50 = {p50:.3f}")
    print(f"  -> top leaf has {mask.sum()} rows (need only 50); true decline rate inside it: {y[mask].mean():.3f}")

Distinct hand_rule_score values: 18 (rest are ties at 0)
depth=2: 4 distinct scores, Precision@50 = 0.600
  -> top leaf has 17816 rows (need only 50); true decline rate inside it: 0.651
depth=4: 16 distinct scores, Precision@50 = 0.680
  -> top leaf has 10442 rows (need only 50); true decline rate inside it: 0.716


In [6]:
(df["hand_rule_score"] > 0).sum() and df["hand_rule_score"].value_counts().head()

,count
hand_rule_score,
0,29983
4590,1
533,1
4556,1
25715,1


In [7]:
print("Distinct hand_rule_score values:", df["hand_rule_score"].nunique())
print("Distinct tree_score values:", len(np.unique(tree_score)))

print("\nTree score distribution (rounded):")
print(pd.Series(tree_score).round(3).value_counts().head(10))

Distinct hand_rule_score values: 18
Distinct tree_score values: 4

Tree score distribution (rounded):
0.612    17816
0.421     9138
0.211     1796
0.007     1250
Name: count, dtype: int64


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.